src/System.Console/src/System/IO/StdInStreamReader.cs

```diff
@@ -0,0 +1,259 @@
+        bool IsExtraBufferEmpty()
+        {
+            if (_startIndex == -1 && _endIndex == -1) return true;
+
+            if (_startIndex > _endIndex) // Everything has been processed;
+            {
+                _startIndex = -1;
+                _endIndex = -1;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary> Reads the buffer in the raw mode. </summary>
+        unsafe void ReadExtraBuffer()
+        {
+            Debug.Assert(IsExtraBufferEmpty());
+
+            byte* bufPtr = stackalloc byte[BytesToBeRead];
+            int result;
+            Interop.CheckIo(result = Interop.Sys.ReadStdinUnbuffered(bufPtr, BytesToBeRead));
+
+            Debug.Assert(result <= BytesToBeRead);
+
+            // Now we need to convert the byte buffer to char buffer.
+            fixed (char* unprocessedBufPtr = _unprocessedBufferToBeRead)
+            {
+                _startIndex = 0;
+                int charCount = CurrentEncoding.GetChars(bufPtr, result, unprocessedBufPtr, _unprocessedBufferToBeRead.Length);
+                _endIndex = charCount - 1;
+            }
+        }
+
+        public override string ReadLine()
+        {
+            //Check if there is anything left in the _unprocessedBufferToBeRead.
+            while (!IsExtraBufferEmpty())
+            {
+                ConsoleKeyInfo keyInfo = ReadKey();
+
+                if (keyInfo.Key == ConsoleKey.Enter)
+                    return _readLineSB.ToString();
+
+                else if (keyInfo.Key == ConsoleKey.Backspace)
+                {
+                    int len = _readLineSB.Length;
+                    if (len > 0)
+                    {
+                        _readLineSB.Length = len - 1;
+                    }
+                }
+
+                else if (keyInfo.KeyChar != '\0') // We ignore all non-printable keys
+                    _readLineSB.Append(keyInfo.KeyChar);
+            }
+
+            string readLineStr = _readLineSB.Append(base.ReadLine()).ToString();
+            _readLineSB.Clear();
+
+            return readLineStr;
+        }
+
+
+        public override int Read()
+        {
+            int x;
+
+            // Convert byte to char.
+            if (IsExtraBufferEmpty())
+            {
+                x = base.Read();
+            }
+            else
+            {
+                x = (int)_unprocessedBufferToBeRead[_startIndex++];
+            }
+
+            return x;
+        }
+
+        internal ConsoleKey GetKeyFromCharValue(ref char x, out bool isShift, out bool isCtrl)
+        {
+            isShift = false;
+            isCtrl = false;
+
+            switch (x)
+            {
+                case '\b':
+                    x = '\0'; // Non-printable character.
+                    return ConsoleKey.Backspace;
+
+                case '\t':
+                    x = '\0'; // Non-printable character.
```

Comment:
Windows and Unix do behave differently here. I was using Console.WriteLine() in my testing so these chars were getting masked, however, I also found that Console.Write() behaves differently for '\0' on windows and Unix. My guess is its because of encoding but I am looking at it more closely.